pkg/provisioners/kafka/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,269 @@
+	defer d.updateLock.Unlock()
+
+	if diff := d.ConfigDiff(config); diff != "" {
+		d.logger.Info("Updating config (-old +new)", zap.String("diff", diff))
+
+		newSubs := make(map[eventingduck.ChannelSubscriberSpec]bool)
+
+		// Subscribe to new subscriptions
+		for _, cc := range config.ChannelConfigs {
+			channelRef := buses.ChannelReference{
+				Name:      cc.Name,
+				Namespace: cc.Namespace,
+			}
+			// Subscribe to new subscriptions
+			for _, sub := range cc.FanoutConfig.Subscriptions {
+				if _, ok := d.kafkaConsumers[channelRef][sub]; ok {
+					// subscribe can be called multiple times for the same subscription,
+					//unsubscribe before we resubscribe
+					err := d.unsubscribe(channelRef, sub)
+					if err != nil {
+						return err
+					}
+				}
+				d.subscribe(channelRef, sub)
+				newSubs[sub] = true
+			}
+		}
+
+		// Unsubscribe and close consumer for any deleted subscriptions
+		for channelRef, subMap := range d.kafkaConsumers {
+			for sub, _ := range subMap {
+				if ok := newSubs[sub]; !ok {
+					d.unsubscribe(channelRef, sub)
+				}
+			}
+		}
+
+		// Update the config so that it can be used for comparison during next sync
+		d.setConfig(config)
+	}
+	return nil
+}
+
+// Run starts the kafka dispatcher's message processing.
+func (d *Dispatcher) Start(stopCh <-chan struct{}) error {
+	if d.kafkaAsyncProducer == nil {
+		return fmt.Errorf("kafkaAsyncProducer is not set")
+	}
+	go func() {
+		for {
+			select {
+			case e := <-d.kafkaAsyncProducer.Errors():
+				d.logger.Warn("Got", zap.Error(e))
+			case s := <-d.kafkaAsyncProducer.Successes():
+				d.logger.Info("Sent", zap.Any("success", s))
+			case <-stopCh:
+				return
+			}
+		}
+	}()
+
+	if d.receiver == nil {
+		return fmt.Errorf("message receiver is not set")
+	}
+
+	d.receiver.Run(stopCh)
+
+	return nil
+}
+
+func (d *Dispatcher) subscribe(channelRef buses.ChannelReference, sub eventingduck.ChannelSubscriberSpec) error {
+
+	d.logger.Info("Subscribing", zap.Any("channelRef", channelRef), zap.Any("subscription", sub))
+
+	topicName := topicName(channelRef)
+
+	group := fmt.Sprintf("%s.%s.%s", controller.Name, sub.Ref.Namespace, sub.Ref.Name)
+	consumerConfig := cluster.NewConfig()
+	consumerConfig.Version = sarama.V1_1_0_0
+	consumer, err := cluster.NewConsumer(d.kafkaBrokers, group, []string{topicName}, consumerConfig)
+	if err != nil {
+		return err
+	}
+
+	channelMap, ok := d.kafkaConsumers[channelRef]
+	if !ok {
+		channelMap = make(map[eventingduck.ChannelSubscriberSpec]*cluster.Consumer)
+		d.kafkaConsumers[channelRef] = channelMap
+	}
+	channelMap[sub] = consumer
+
+	go func() {
+		for {
+			msg, more := <-consumer.Messages()
+			if more {
+				d.logger.Info("Dispatching a message for subscription", zap.Any("channelRef", channelRef), zap.Any("subscription", sub))
+				message := fromKafkaMessage(msg)
+				err := d.dispatchMessage(message, sub)
+				if err != nil {
+					d.logger.Warn("Got error trying to dispatch message", zap.Error(err))
```

Comment:
getting:

```
{"level":"info","ts":1541648221.1524541,"caller":"buses/message_receiver.go:110","msg":"Received request for my-k-channel.default.channels.cluster.local"}
{"level":"info","ts":1541648221.1623485,"caller":"dispatcher/dispatcher.go:166","msg":"Dispatching a message for subscription","channelRef":{"Namespace":"default","Name":"my-k-channel"},"subscription":{"ref":{"namespace":"default","name":"testevents-subscription","uid":"4545a607-e307-11e8-83bd-5254009e6bd8"},"subscriberURI":"http://message-dumper.default.svc.cluster.local/"}}
{"level":"info","ts":1541648221.1625435,"caller":"buses/message_dispatcher.go:93","msg":"Dispatching message to http://message-dumper.default.svc.cluster.local/"}
{"level":"info","ts":1541648221.162761,"caller":"dispatcher/dispatcher.go:166","msg":"Dispatching a message for subscription","channelRef":{"Namespace":"default","Name":"my-k-channel"},"subscription":{"ref":{"namespace":"default","name":"testevents-kafka-subscription","uid":"afa5d34a-e2ab-11e8-83bd-5254009e6bd8"},"subscriberURI":"http://message-dumpy.default.svc.cluster.local/"}}
{"level":"info","ts":1541648221.1627984,"caller":"buses/message_dispatcher.go:93","msg":"Dispatching message to http://message-dumpy.default.svc.cluster.local/"}
{"level":"warn","ts":1541648221.1700735,"caller":"dispatcher/dispatcher.go:170","msg":"Got error trying to dispatch message","error":"Unable to complete request Post http://message-dumper.default.svc.cluster.local/: dial tcp: lookup message-dumper.default.svc.cluster.local on 172.30.0.2:53: no such host"}

```

Reply:
Ok, that I sorted out. and i do receive messages from the k8s event source :tada: :balloon: